logo.js

```diff
@@ -924,84 +924,38 @@ function LogoInterpreter(turtle, stream, savehook)
     self.forceBye = true;
   };
 
-  self.run = function(string, options) {
-    if (self._pendingPromise) {
-      var runner = self.run.bind(self, string, options);
-      return self._queueCall(runner);
-    }
-    options = Object(options);
-    // Parse it
-    var atoms = parse(string);
-
-    if (self.turtle) { self.turtle.begin(); }
-    // And execute it!
-    var promise = self.execute(atoms, options).then(function (result) {
-      if (self.turtle) {
-        self.turtle.end();
-      }
-      return result;
-    }, function (err) {
-      if (self.turtle) {
-        self.turtle.end();
-      }
-      if (err && (err.special == "bye" || err.special == "stop")) {
-        return undefined;
-      }
-      throw err;
-    });
-    this._setPendingPromise(promise);
-    self._pendingPromise = promise;
-    return promise;
-  };
+  function noop() {}
+  var lastRun = Promise.resolve();
 
-  //----------------------------------------------------------------------
-  // Helpers to ensure calls to run() are serialized:
-  //----------------------------------------------------------------------
+  self.run = function(string, options) {
+    var promise = lastRun.then(function() {
```

Comment:
I think the code I wrote did something like:

```javascript
lastRun = null;
self.run = function (expr) {
  if (lastRun) {
    return lastRun.then(self.run.bind(this, expr));
  }
  lastRun = ...;
  return lastRun;
};
```

This meant when you had 10 calls to .run() they all watched `lastRun` and all attempted to execute themselves, even though one would run and the other 9 would only chain themselves on the new lastRun.  This looks the same... but now that I'm remembering it, that also caused errors because routines weren't actually called in order.  Is that not what you are seeing with this code?  The logic seems... similar...?